src/TrackGenerator.cpp

```diff
@@ -307,42 +313,14 @@ void TrackGenerator::calculateFSRVolumes() {
     log_printf(ERROR, "Unable to calculate the FSR volumes since the FSR "
                "volumes array has not yet been allocated");
 
-
   /* Reset FSR volumes to zero */
   int num_FSRs = _geometry->getNumFSRs();
-  memset(_FSR_volumes, 0., num_FSRs*sizeof(FP_PRECISION));
-
-#pragma omp parallel
-  {
-    int azim_index, fsr_id;
-    segment* curr_segment;
-    FP_PRECISION volume;
-
-    /* Calculate each FSR's "volume" by accumulating the total length of *
-     * all Track segments multipled by the Track "widths" for each FSR.  */
-    for (int i=0; i < _num_azim_2; i++) {
-#pragma omp for
-      for (int j=0; j < _num_tracks[i]; j++) {
-
-        azim_index = _tracks[i][j].getAzimAngleIndex();
-
-        for (int s=0; s < _tracks[i][j].getNumSegments(); s++) {
-          curr_segment = _tracks[i][j].getSegment(s);
-          volume = curr_segment->_length * _quadrature->getAzimWeight(i)
-            * _quadrature->getAzimSpacing(i);
-          fsr_id = curr_segment->_region_id;
-
-          /* Set FSR mutual exclusion lock */
-          omp_set_lock(&_FSR_locks[fsr_id]);
-
-          _FSR_volumes[fsr_id] += volume;
+  if (_FSR_volumes != NULL)
+    memset(_FSR_volumes, 0., num_FSRs*sizeof(FP_PRECISION));
 
-          /* Release FSR mutual exclusion lock */
-          omp_unset_lock(&_FSR_locks[fsr_id]);
-        }
-      }
-    }
-  }
+  /* Create volume calculator and calculate new FSR volumes */
+  VolumeCalculator volume_calculator(this);
```

Comment:
Just to be clear, the idea is that this will work as is for a variety of tracking schemes including all combinations of explicit/OTF 2D/3D?